README
Clutter version 0.420
=====================

Perl bindings to the Clutter OpenGL canvas.  Clutter aims to give a nice
and easy GObject based API for creating fast, mainly 2B single window
stylised applications, such as media box UI, presentations, kiosk style
applications, etc.  Clutter uses OpenGL for rendering.

The Clutter Perl bindings allow you to write canvas-based applications in
a Perlish and object-oriented way, freeing you from having to care about
casting and memory management of the original Clutter C API, and yet
remaining close to its spirit.

You can find out more about Clutter at http://www.clutter-project.org .

INSTALLATION
------------

To install this module type the following:

  perl Makefile.PL
  make
  make test
  make install

To avoid installing to a system directory you can change the installation
prefix at Makefile.PL time with

  perl Makefile.PL PREFIX=/some/other/place

This will install the module to the subdirectory lib/perl5 under the given
prefix.  If this is not already in Perl's include path, you'll need to tell
perl how to the to this library directory so you can use it.

DEPENDENCIES
------------

This module requires these other modules and libraries:

  perl >= 5.8.0
  Glib >= 1.120 (Perl module)
  Gtk2 >= 1.120 (Perl module)
  Clutter >= 0.4.2 (C library)
  ClutterGst >= 0.4.0 (C library)
  ClutterGtk >= 0.4.0 (C library)
  ClutterCairo >= 0.4.0 (C library)

In order to build it from source, you'll also need:

  ExtUtils::Depends >= 0.2
  ExtUtils::PkgConfig >= 1.03

BUG REPORTS
-----------

Please report bugs to the clutter product in OpenedHand's Bugzilla:

  http://bugzilla.gnome.org/enter_bug.cgi?product=Clutter

COPYRIGHT AND LICENSE
---------------------

Copyright (C) 2006-2007  OpenedHand Ltd.  (see the file AUTHORS)

This library is free software; you can redistribute it and/or modify it under
the terms of the GNU Library General Public License as published by the Free
Software Foundation; either version 2.1 of the License, or (at your option)
any later version

This library is distributed in the hope that it will be useful, but WITHOUT
ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
FOR A PARTICULAR PURPOSE.  See the GNU Library General Public License for more
details.

You should have received a copy of the GNU Library General Public License along
with this library; if not, write to the Free Software Foundation, Inc., 59
Temple Place - Suite 330, Boston, MA  02111-1307  USA.


NEWS
Clutter 0.420
=============
* Fix methods collisions in Clutter::Actor, Clutter::Entry and Clutter::Media
* Require the new stable release of Clutter

Clutter 0.410
=============
* Fix Clutter::Actor::QUERY_COORDS to return an array
* Add a test for subclassing Clutter::Actor
* Documentation fixes in Clutter::Units and Clutter::Layout
* Allow a pure Perl implementation of Clutter::Layout::TUNE_REQUEST
* Bind more Clutter::Effect API
* Require the new stable release of Clutter

Clutter 0.402
=============
* Make subclassing Clutter::Box more perlish
* Add Clutter->SUPPORTED_MODULES(), to check the supported
  sub-modules against which Clutter was compiled
* Make Clutter::TestHelper skip tests if no support for the
  needed module was found

Clutter 0.401
=============
* Build fixes for packaging
* Add optional bindings for the Gtk2 integration
* Add a Test::More wrapper module and update the test suite

Clutter 0.400
=============
* Resync with upstream API
* Include autogenerated key symbols as Clutter::Keysyms
* Allow overriding Clutter::Actor virtual methods
* Add optional bindings for the GStreamer integration
* Add optional bindings for the Cairo integration
* Register a sink function for every initially unowned class
* Add more tests


Clutter.pm
#
# Copyright (c) 2006  OpenedHand Ltd. (see the file AUTHORS)
#
# This library is free software; you can redistribute it and/or
# modify it under the terms of the GNU Library General Public
# License as published by the Free Software Foundation; either
# version 2 of the License, or (at your option) any later version.
#
# This library is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# Library General Public License for more details.
#
# You should have received a copy of the GNU Library General Public
# License along with this library; if not, write to the 
# Free Software Foundation, Inc., 59 Temple Place - Suite 330, 
# Boston, MA  02111-1307  USA.

package Clutter;

use 5.008;
use strict;
use warnings;

use Glib;
use Gtk2;

require DynaLoader;

our @ISA = qw( DynaLoader );

# the version scheme is:
#   CLUTTER_MAJOR
#   dot
#   CLUTTER_MINOR * 100 + CLUTTER_MICRO * 10 + bindings release
#
# this scheme allocates enough space for ten releases
# of the bindings for each point release of libclutter,
# which should be enough even in case of brown paper
# bag releases. -- ebassi
our $VERSION = '0.420';

sub import {
    my $class = shift;

    # Clutter::Gst->init() is a wrapper around the GStreamer init and
    # Clutter init calls, so using gst-init or init is equivalent for
    # us. in case Clutter::Gst wasn't compiled in, Clutter::Gst->init()
    # will expand to Clutter->init() anyway.

    # Clutter::Threads->init() must be called before calling Clutter->init(),
    # but we don't want to force the order of the options passed, so we store
    # the choices and call everything in the correct order later.

    my $init = 0;
    my $threads_init = 0;

    foreach (@_) {
        if    (/^[-:]?init$/)        { $init = 1;           }
        elsif (/^[-:]?gst-init$/)    { $init = 2;           }
        elsif (/^[-:]?threas-init$/) { $threads_init = 0;   }
        else                         { $class->VERSION($_); }
    }

    Clutter::Threads->init() if $threads_init;
    Clutter->init()          if $init == 1;
    Clutter::Gst->init()     if $init == 2;
}

sub dl_load_flags { $^O eq 'darwin' ? 0x00 : 0x01 }

require XSLoader;
XSLoader::load('Clutter', $VERSION);

# Preloaded methods go here

package Clutter::Color;

use overload
    '==' => \&Clutter::Color::equal,
    '+' => \&Clutter::Color::add,
    '-' => \&Clutter::Color::subtract,
    fallback => 1;

package Clutter::Knot;

use overload
    '==' => \&Clutter::Knot::equal,
    fallback => 1;

package Clutter::Vertex;

use overload
    '==' => \&Clutter::Vertex::equal,
    fallback => 1;

package Clutter;

1;

__END__

=pod

=head1 NAME

Clutter - Simple GL-based canvas library

=head1 SYNOPSIS

  use Clutter qw( :init );
  
  # create the main stage
  my $stage = Clutter::Stage->get_default();
  $stage->set_color(Clutter::Color->parse('DarkSlateGray'));
  $stage->signal_connect('key-press-event' => sub { Clutter->main_quit() });
  $stage->set_size(800, 600);
  
  # add an actor and place it right in the middle
  my $label = Clutter::Label->new("Sans 30", "Hello, Clutter!");
  $label->set_color(Clutter::Color->new(0xff, 0xcc, 0xcc, 0xdd));
  $label->set_position(($stage->get_width()  - $label->get_width())  / 2,
                       ($stage->get_height() - $label->get_height()) / 2);
  $stage->add($label);

  $stage->show_all();
  
  Clutter->main();
  
  0;

=head1 DESCRIPTION

Clutter is a GObject based library for creating fast, visually rich
graphical user interfaces.  It is intended for creating single window
heavily stylised applications such as media box ui's, presentations or
kiosk style programs in preference to regular 'desktop' style
applications.

Clutter's underlying graphics rendering is OpenGL (version 1.2+)
based.  The clutter API is intended to be easy to use, attempting to
hide many of the GL complexities.  It targets mainly 2D based graphics
and is definetly not intended to be a general interface for all OpenGL
functionality.

As well as OpenGL Clutter depends on and uses Glib, Glib::Object,
Gtk2::Pango, Gtk2::Gdk::Pixbuf and GStreamer.

For more informations about Clutter, visit:

  http://www.clutter-project.org

You can also subscribe to the Clutter mailing list by sending a
blank message E<lt>clutter+subscribe AT o-hand.comE<gt>, then follow
the instructions in resulting reply.

=head1 DIFFERENCES FROM C API

In order to feel more Perl-ish, the Clutter API has been slightly
changed for the Perl bindings.

=over 4

=item ClutterCloneTexture =E<gt> Clutter::Texture::Clone

The C<ClutterCloneTexture> has been moved under the L<Clutter::Texture>
package name, to reinforce the inheritance.

=item ClutterCairo =E<gt> Clutter::Texture::Cairo

As above, the name has been changed to reinforce the inheritance.

=back

=head1 AUTHOR

Emmanuele Bassi E<lt>ebassi (AT) openedhand (DOT) comE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2006  OpenedHand Ltd.

This module is free software; you can redistribute it and/or
modify it under the terms of the GNU Lesser General Public
License as published by the Free Software Foundation, version 2.1;
or, at your option, under the terms of The Artistic License.

This module is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

You should have received a copy of the GNU Library General Public
License along with this module; if not, write to the 
Free Software Foundation, Inc., 59 Temple Place - Suite 330, 
Boston, MA  02111-1307  USA.

For the terms of The Artistic License, see L<perlartistic>.

=cut


tools/TestHelper.pm
package Clutter::TestHelper;

use strict;
use warnings;
use Test::More;
use Carp;

our $VERSION = '0.01';

sub import
{
    shift;
    my %opts = (@_);

    plan skip_all => $opts{skip_all}
        if $opts{skip_all};

    croak "test must be provided at import"
        unless exists $opts{tests};

    if ($opts{at_least_version}) {
        my ($rmajor, $rminor, $rmicro, $text) = @{$opts{at_least_version}};

        plan skip_all => $text
            unless Clutter->CHECK_VERSION($rmajor, $rminor, $rmicro);
    }

    if ($opts{sub_module}) {
        my @modules = Clutter->SUPPORTED_MODULES();
        my $no_skip = 0;

        foreach my $m (@modules) {
            $no_skip = 1 if $opts{sub_module} eq $m;
        }

        plan skip_all => "no support for $opts{sub_module}"
            unless $no_skip;
    }

    if (!Clutter->init()) {
        plan skip_all => 'Clutter->init failed';
    }
    else {
        plan tests => $opts{tests};
    }

    # forcibly turn on warnings and strictness in the caller
    $^W = 1;
    @_ = ();
    goto &strict::import;
}

package main;

# we obviously need Clutter
use Clutter;

# and Test::More
use Test::More;

# enforce the use of these constants in the tests
use Glib qw( TRUE FALSE );

1;
__END__

=head1 NAME

Clutter::TestHelper - Simple wrapper module for testing Clutter

=head1 SYNOPSIS

  use Clutter::TestHelper tests => 10;

=head1 DESCRIPTION

A simple wrapper module around L<Test::More> that makes Clutter test suite
easier to write. When importing the module, L<Clutter> is also imported and
initialised. In case the initialisation process fails, likely to happen if
the C<DISPLAY> environment variable is not set, all the planned tests are
just skipped.

=head1 OPTIONS

=over

=item tests => $number

The number of tests to be performed

=item at_least_version => [ $major, $minor, $micro, $reason ]

A reference to a list that is checked with Clutter->CHECK_VERSION

=item skip_all

Simply skip all tests

=item sub_module => $module_name

Skip all tests if Clutter was not compiled against I<module_name>.

=back

=head1 AUTHORS

Emmanuele Bassi  E<lt>ebassi (AT) openedhand (DOT) comE<gt>

This module heavily borrows from the equivalent L<Gtk2::TestHelper> module
written by the Gtk2-Perl team.

=head1 COPYRIGHT

Copyright (C) 2007  OpenedHand, Ltd.

Released under the terms of Clutter itself.

=cut


Makefile.PL
use 5.008;

use strict;
use warnings;

use Cwd;
use File::Spec;
use ExtUtils::MakeMaker;

# minimum required version of dependancies we need to build
our %build_reqs = (
    'perl-ExtUtils-Depends'   => '0.2',
    'perl-ExtUtils-PkgConfig' => '1.03',
    'perl-Glib'               => '1.140', # for Glib::MakeHelper
    'perl-Gtk2'               => '1.140',
    'perl-Cairo'              => '1.00',
    'Clutter'                 => '0.4.2',
);

# Writing a fake Makefile ensures that CPAN will pick up the correct
# dependencies and install them.
unless (eval "use ExtUtils::Depends '$build_reqs{'perl-ExtUtils-Depends'}';"
           . "use ExtUtils::PkgConfig '$build_reqs{'perl-ExtUtils-PkgConfig'}';"
           # just seeing if Glib is available isn't enough, make sure
           # it's recent enough, too 
           . "use Glib '$build_reqs{'perl-Glib'}';"
           . "use Gtk2 '$build_reqs{'perl-Gtk2'}';"
           . "use Glib::MakeHelper;"
           # we need Gtk2::CodeGen because we use GtkObject in ClutterGtk
           . "use Gtk2::CodeGen;"
           . "1") {
   warn "$@\n";
   WriteMakefile(
         PREREQ_FATAL => 1,
         PREREQ_PM    => {
            'Glib'                => $build_reqs{'perl-Glib'},
            'Gtk2'                => $build_reqs{'perl-Gtk2'},
            'ExtUtils::Depends'   => $build_reqs{'perl-ExtUtils-Depends'},
            'ExtUtils::PkgConfig' => $build_reqs{'perl-ExtUtils-PkgConfig'},
         },
   );
   exit 1; # not reached
}

mkdir 'build', 0777;

our @xs_files = ();

# Base module
our %pkgcfg =
    ExtUtils::PkgConfig->find('clutter-0.4 >= ' . $build_reqs{Clutter});

my @names = Glib::MakeHelper->read_source_list_file ('xs_files');
print "Loaded " . scalar(@names) . " xs files for Clutter\n";
push @xs_files, @names;

our $module_cflags = $pkgcfg{cflags};
our $module_libs   = $pkgcfg{libs};

# Integration libraries: Clutter::Gst
our $have_clutter_gst = 0;
eval {
    my %gstpkgcfg = ExtUtils::PkgConfig->find('clutter-gst-0.4');

    $module_cflags .= ' ' . $gstpkgcfg{cflags} . '-DCLUTTERPERL_GST';
    $module_libs   .= ' ' . $gstpkgcfg{libs};

    $have_clutter_gst = 1;

    my @names = Glib::MakeHelper->read_source_list_file ('xs_files_gst');
    print "Loaded " . scalar(@names) . " xs files for Clutter::Gst\n";
    push @xs_files, @names;
};
if ($@) {
    warn("*** No clutter-gst found, skipping Clutter::Gst\n");
}

# Integration libraries: Clutter::Texture::Cairo
our $have_clutter_cairo = 0;
eval {
    my %cairopkgcfg = ExtUtils::PkgConfig->find('clutter-cairo-0.4');

    $module_cflags .= ' ' . $cairopkgcfg{cflags} . '-DCLUTTERPERL_CAIRO';
    $module_libs   .= ' ' . $cairopkgcfg{libs};

    $have_clutter_cairo = 1;

    my @names = Glib::MakeHelper->read_source_list_file ('xs_files_cairo');
    print "Loaded " . scalar(@names) . " xs files for Clutter::Cairo\n";
    push @xs_files, @names;
};
if ($@) {
    warn("*** No clutter-cairo found, skipping Clutter::Cairo\n");
}

# Integration libraries: Gtk2::ClutterEmbed
our $have_clutter_gtk = 0;
eval {
    my %gtkpkgcfg = ExtUtils::PkgConfig->find('clutter-gtk-0.4');

    $module_cflags .= ' ' . $gtkpkgcfg{cflags} . '-DCLUTTERPERL_GTK';
    $module_libs   .= ' ' . $gtkpkgcfg{libs};

    $have_clutter_gtk = 1;

    my @names = Glib::MakeHelper->read_source_list_file ('xs_files_gtk');
    print "Loaded " . scalar(@names) . " xs files for Clutter::Gtk\n";
    push @xs_files, @names;
};
if ($@) {
    warn("*** No clutter-gtk found, skipping Clutter::Gtk\n");
}


our %pm_files = (
    'Clutter.pm' => '$(INST_LIBDIR)/Clutter.pm',

    # Keysyms.pm is generated by Makefile.PL
    'build/Keysyms.pm' => '$(INST_LIBDIR)/Clutter/Keysyms.pm',

    # TestHelper is just a wrapper around Test::More
    'tools/TestHelper.pm' => '$(INST_LIBDIR)/Clutter/TestHelper.pm',
);

our %pod_files = (
    'Clutter.pm'   => '$(INST_MAN3DIR)/Clutter.$(MAN3EXT)',

    # Keysyms.pm is generated by Makefile.PL
    'build/Keysyms.pm' => '$(INST_MAN3DIR)/Clutter::Keysyms.$(MAN3EXT)',
    
    Glib::MakeHelper->do_pod_files(@xs_files),
);

# typemaps generation files
my @input_maps = qw/ maps /;
push @input_maps, 'maps-gst'   if $have_clutter_gst   == 1;
push @input_maps, 'maps-cairo' if $have_clutter_cairo == 1;
push @input_maps, 'maps-gtk'   if $have_clutter_gtk   == 1;

# autogeneration
Gtk2::CodeGen->parse_maps('clutterperl', input => \@input_maps);
Gtk2::CodeGen->write_boot(xs_files => \@xs_files);

# this one turns the long list of #defines in clutter-keysyms.h into a
# large hash of key names and values, %Clutter::Gdk::Keysyms.
system "$^X tools/genkeysyms.pl >\"build/Keysyms.pm\"";

my @deps = qw/ Glib /;
push @deps, 'GStreamer' if $have_clutter_gst   == 1;
push @deps, 'Cairo'     if $have_clutter_cairo == 1;
push @deps, 'Gtk2';     # default, because we need Gtk2::Gdk::Pixbuf

my $cwd = cwd();

our $clutter = ExtUtils::Depends->new('Clutter', @deps);
$clutter->set_inc($module_cflags . ' -I./build ');
$clutter->set_libs($module_libs);
$clutter->add_xs(@xs_files);
$clutter->add_pm(%pm_files);
$clutter->add_typemaps(map { File::Spec->catfile($cwd, $_) }
                           'build/clutterperl.typemap');
$clutter->install(qw(clutterperl.h build/clutterperl-autogen.h));
$clutter->save_config('build/IFiles.pm');

WriteMakefile(
    NAME          => 'Clutter',
    VERSION_FROM  => 'Clutter.pm', # finds $VERSION
    PREREQ_PM     => {}, # e.g., Module::Name => 1.1
    ABSTRACT_FROM => 'Clutter.pm', # retrieve abstract from module
    XSPROTOARG    => '-noprototypes',
    MAN3PODS	  => \%pod_files,
    $clutter->get_makefile_vars(),
);


=pod
#=unstable

print "
WARNING WARNING WARNING WARNING WARNING WARNING WARNING WARNING WARNING

This is an unstable development release of Clutter.  The API is
not frozen and things are subject to change at any time.  Report any
bugs to clutter AT openedhand DOT com as soon as possible.

WARNING WARNING WARNING WARNING WARNING WARNING WARNING WARNING WARNING
";

=cut

sub MY::postamble
{
    return Glib::MakeHelper->postamble_clean()
         . Glib::MakeHelper->postamble_docs_full(
            DEPENDS => $clutter,
            DOCTYPES => 'doctypes',
            COPYRIGHT_FROM => 'copyright.pod',
           )
         . Glib::MakeHelper->postamble_rpms(
            'CLUTTER' => $build_reqs{'Clutter'},
            'PERL_EXTUTILS_DEPENDS' => $build_reqs{'perl-ExtUtils-Depends'},
            'PERL_EXTUTILS_PKGCONFIG' => $build_reqs{'perl-ExtUtils-PkgConfig'},
            'PERL_GLIB' => $build_reqs{'perl-Glib'},
            'PERL_GTK2' => $build_reqs{'perl-Gtk2'},
           );
}
